app/javascript/components/MapAndButtons.jsx

```diff
@@ -9,8 +9,13 @@ import MapView from './MapView';
 import NotificationList from './NotificationList';
 import SearchFilterAccordion from './SearchFilterAccordion';
 import SplashPage from './SplashPage';
+import Legend from './legend'
 class MapAndButtons extends React.Component {
 
+  create_legend_set(locations){
+    return [...new Set(locations.map(loc => loc.icon_url.split("?")[0] + ":" + loc.icon_name + ":" + loc.floor_id))] 
```

Comment:
filter on floor_id here, and map to create a shape for properties to pass to the legend.. perhaps an array of `{icon_url, icon_name}` objects?

Reply:
something along these lines, 

```
..filter on floor_id to create your list of persistent locations and searched locations

let persistent_icons = persistent_locations.map(loc => {icon_url: loc.icon_url, icon_name: loc.icon_name})

let icons = searched_locations.map(loc => {icon_url: loc.icon_url, icon_name: loc.icon_name})

return new Set(...persistent_icons, ...icons)
```